processing/src/main/java/org/apache/druid/query/topn/types/StringTopNColumnAggregatesProcessor.java

```diff
@@ -172,6 +180,54 @@ private long scanAndAggregateWithCardinalityKnown(
     return processedRows;
   }
 
+  /**
+   * this method uses hash table to store aggregate values ({@link #aggregatesStore}) and is geared towards selectors
+   * which have a dictionary, but those dictionary ids do not necessarily uniquely map to the value of
+   * {@link DimensionSelector#lookupName(int)}. This method keeps a cache of dictionary id to aggregators, which it uses
+   * to skip name lookup if that id has already been encountered, avoiding decoding utf8 bytes, or worse, recomputing
+   * an expression virtual column or some lookup value.
+   *
+   */
+  private long scanAndAggregateWithDictionary(
+      TopNQuery query,
+      Cursor cursor,
+      DimensionSelector selector
+  )
+  {
+    Int2ObjectOpenHashMap<Aggregator[]> aggsCache = new Int2ObjectOpenHashMap<>();
```

Comment:
With the test for `hasDictionary` we could perhaps make this

```
Aggregator[][] aggsCache = new Aggregator[selector.getValueCardinality()][];
```

?

Reply:
yeah, that should work :+1: